tests/h/services/annotation_metadata_test.py

```diff
@@ -0,0 +1,45 @@
+from unittest.mock import sentinel
+
+import pytest
+
+from h.models import AnnotationMetadata
+from h.services.annotation_metadata import AnnotationMetadataService, factory
+
+
+class TestAnnotationMetadataService:
+    def test_it(self, svc, factories, decrypt_jwe_dict, db_session):
+        decrypt_jwe_dict.side_effect = [{"some": "data"}, {"new": "data"}]
+        anno = factories.Annotation()
+        anno_slim = factories.AnnotationSlim(annotation=anno)
+        db_session.flush()
+
+        svc.set_annotation_metadata_from_jwe(anno, sentinel.jwe)
+
+        decrypt_jwe_dict.assert_called_once_with("TEST_JWE_SECRET_LMS", sentinel.jwe)
+        anno_metadata = (
+            db_session.query(AnnotationMetadata)
+            .filter_by(annotation_id=anno_slim.id)
+            .one()
+        )
+        assert anno_metadata.data == {"some": "data"}
+
+        # Second call updates the existing record
+        svc.set_annotation_metadata_from_jwe(anno, sentinel.jwe)
+        db_session.refresh(anno_metadata)
+        assert anno_metadata.data == {"new": "data"}
+
+    def test_factory(self, pyramid_request):
+        pyramid_request.registry.settings["jwe_secret_lms"] = "TEST_JWE_SECRET_LMS"
+
+        svc = factory(sentinel.context, pyramid_request)
+
+        assert isinstance(svc, AnnotationMetadataService)
```

Comment:
This could be passing the wrong arguments to `AnnotationMetadataService` and the test would still pass. For example this wouldn't break the test:

```diff
diff --git a/h/services/annotation_metadata.py b/h/services/annotation_metadata.py
index fb1c9769c..3907f2366 100644
--- a/h/services/annotation_metadata.py
+++ b/h/services/annotation_metadata.py
@@ -46,4 +46,4 @@ class AnnotationMetadataService:
 
 
 def factory(_context, request) -> AnnotationMetadataService:
-    return AnnotationMetadataService(db=request.db)
+    return AnnotationMetadataService(db="db")
```

I reckon it's actually better to use a mock here because you can test that the `db_session` was extracted from the request properly (`request.db`) and that `AnnotationMetadataService()` was called with the right values. If more arguments are added to `AnnotationMetadataService()` in future (e.g. config settings from `request.registry.settings`, other services from `request.find_service()`, etc) the benefit grows.

```diff
diff --git a/h/services/annotation_metadata.py b/h/services/annotation_metadata.py
index fb1c9769c..15c76da69 100644
--- a/tests/h/services/annotation_metadata_test.py
+++ b/tests/h/services/annotation_metadata_test.py
@@ -28,10 +28,14 @@ class TestAnnotationMetadataService:
         db_session.refresh(anno_metadata)
         assert anno_metadata.data == {"new": "data"}
 
-    def test_factory(self, pyramid_request):
+    def test_factory(self, AnnotationMetadataService, db_session, pyramid_request):
         svc = factory(sentinel.context, pyramid_request)
 
-        assert isinstance(svc, AnnotationMetadataService)
+        AnnotationMetadataService.assert_called_once_with(db=db_session)
+
+    @pytest.fixture
+    def AnnotationMetadataService(self, patch):
+        return patch("h.services.annotation_metadata.AnnotationMetadataService")
 
     @pytest.fixture
     def decrypt_jwe_dict(self, patch):
```

Reply:
Yeah, that a much better test :+1: